ntoskrnl/io/debug.c

```diff
@@ -0,0 +1,561 @@
+/*
+ * PROJECT:         ReactOS Kernel
+ * COPYRIGHT:       GPL - See COPYING in the top level directory
+ * FILE:            ntoskrnl/io/debug.c
+ * PURPOSE:         Useful functions for debugging IO and PNP managers
+ * PROGRAMMERS:     
+ */
+
+#include <ntoskrnl.h>
+#include "pnpio.h"
+
+#define NDEBUG
+#include <debug.h>
+
+/* GLOBALS ********************************************************************/
+
+extern PDEVICE_NODE IopRootDeviceNode;
+
+/* FUNCTIONS ******************************************************************/
+
+/* CmResource */
+
+/* PipDumpCmResourceDescriptor() displays information about a Cm Descriptor 
+
+   DebugLevel: 0 - always dump
+               1 - dump if not defined NDEBUG
+*/
+VOID
+NTAPI
+PipDumpCmResourceDescriptor(
+    _In_ PCM_PARTIAL_RESOURCE_DESCRIPTOR Descriptor,
+    _In_ ULONG DebugLevel)
+{
+    PAGED_CODE();
+
+    if (DebugLevel != 0)
+    {
+#ifdef NDEBUG
+        return;
+#endif
+    }
+
+    if (Descriptor == NULL)
+    {
+        DPRINT1("Dump CmDescriptor: Descriptor == NULL\n");
```

Comment:
Something to be thought for all the functions in this file (a suggestion):
since "DPRINT1" always appends the full file path + line of the file before the strings, if you think this clutters the debug log, you can use instead DbgPrint().

Reply:
When debugging on a real computer in "screen" mode, you are limited to 80 characters on one line. Therefore, for such cases, I did a separate buid RoS in which I globally cut all the lines debug out in sdk\include\reactos\debug.h. 
``` c
#ifndef __FILENAME__
  #define __FILENAME__ (strrchr(__FILE__, '/') ? strrchr(__FILE__, '/') + 1 : __FILE__)
#endif
```
``` c
      #if DBG_FULL_PATH
        #define DPRINT1(fmt, ...) do { \
            if (DbgPrint("(%s:%d) " fmt, __RELFILE__, __LINE__, ##__VA_ARGS__))  \
                DbgPrint("(%s:%d) DbgPrint() failed!\n", __RELFILE__, __LINE__); \
        } while (0)
      #else
        #define DPRINT1(fmt, ...) do { \
            if (DbgPrint("(%s:%d) " fmt, __FILENAME__, __LINE__, ##__VA_ARGS__))  \
                DbgPrint("(%s:%d) DbgPrint() failed!\n", __FILENAME__, __LINE__); \
        } while (0)
      #endif
```
And also shortened all the long debugging messages that were common.

Therefore, ideally, it would be nice to have a multivariate DPRINT that would print different messages for different purposes. In general, I would replace DPRINT1 with DPRINT0, and DPRINT1 ... DPRINTn could be used for this, for example.

I think this is a separate "global" topic for the new PR. Which will be relevant to simplify debugging on real computers. It is good for a virtual machine to have detailed information; you are not limited to a long line here. The size of the log does not scare me, after the logs from uniata, I’m not afraid of anything :)